misc/version_updater.py

```diff
@@ -0,0 +1,325 @@
+}
+
+
+def check_correct_tool_version(filename: Path, raw_regexes: RawRegexes, version: str) -> List[str]:
+    try:
+        regexes = compile_regexes(raw_regexes, version)
+    except re.error:
+        raise ValueError(f"Failed to compile regexes for file `{filename}`")
+    have_matched = {regex[0].pattern: False for regex in regexes}
+    errors = []
+
+    for line_nu, line in enumerate(filename.read_text().splitlines()):
+        for regex, expected_line in regexes:
+            res = regex.search(line)
+            if res is not None:
+                have_matched[regex.pattern] = True
+                matched_part = res.group(0)
+                if matched_part != expected_line:
+                    effective_line = line_nu + 1
+                    effective_col_start = res.start(0)
+                    errors.append(
+                        f"{filename}:{effective_line}:{effective_col_start}: Invalid version in file got `{matched_part}` but expected `{expected_line}`"
+                    )
+
+    match_errors = does_every_regex_where_used(filename, have_matched)
+    errors.extend(match_errors)
+    return errors
+
+
+def update_tool_version(filename: Path, raw_regexes: RawRegexes, version: str) -> List[str]:
+    try:
+        regexes = compile_regexes(raw_regexes, version)
+    except re.error:
+        raise ValueError(f"Failed to compile regexes for file `{filename}`")
+    have_matched = {regex[0].pattern: False for regex in regexes}
+
+    with FileInput(filename, inplace=True) as f:
+        for line in f:
+            for regex, replaced_line in regexes:
+                line, substitution_count = regex.subn(replaced_line, line, count=1)
+                if substitution_count >= 1:
+                    have_matched[regex.pattern] = True
+                line = line
+            print(line, end="")
+
+    return does_every_regex_where_used(filename, have_matched)
+
+
+def compile_regexes(regexes: RawRegexes, version: str) -> List[Tuple[re.Pattern[str], str]]:
+    return [regex.compile(version) for regex in regexes]
+
+
+def does_every_regex_where_used(filename: Path, have_matched: Dict[str, bool]) -> List[str]:
+    errors = []
+    for pattern, matched in have_matched.items():
+        if not matched:
+            errors.append(f"{filename}: Regex `{pattern}` found no matches")
+
+    return errors
+
+
+def check_tool(tool: Tool, version: str, update: bool) -> Dict[Path, List[str]]:
+    errors = {}
+    for filename, tools_in_file in FILES_WITH_VERSION_INFO.items():
+        regexes = tools_in_file.get(tool, None)
+        if regexes is None:
+            continue
+
+        if update:
+            possible_errors = update_tool_version(filename, regexes, version)
+        else:
+            possible_errors = check_correct_tool_version(filename, regexes, version)
+
+        if len(possible_errors) > 0:
+            errors[filename] = possible_errors
```

Comment:
I would simplify this a little bit by receiving the check or the update function (same arguments) instead of the `update` boolean. 

```suggestion
        file_errors = check_or_update(filename, regexes, version)

        if file_errors:
            errors[filename] = file_errors
```

Reply:
I don't know why the comments/suggestion in my review are not well sorted :(

Reply:
done